File: src/signals/fear-greed.ts
/**
 * Fear & Greed Index fetcher (Alternative.me).
 *
 * The index is computed from:
 *  - Volatility (25%)
 *  - Market momentum / volume (25%)
 *  - Social media (15%)
 *  - Surveys (15%)
 *  - Bitcoin dominance (10%)
 *  - Google Trends (10%)
 *
 * We fetch both current and yesterday's value to detect direction of movement
 * (a falling index that's already at 20 is different from one still in freefall).
 *
 * Refreshed every 30 minutes — the index only updates daily, but we cache
 * appropriately and expose the movement delta.
 */

import { log } from '../storage/database.js';
import type { MarketContext, MarketPhase } from '../types.js';

interface FGIDataPoint {
  value: string;
  value_classification: string;
  timestamp: string;
}

interface FGIResponse {
  data: FGIDataPoint[];
}

export interface FearGreedReading {
  index: number;        // 0-100
  label: string;        // "Extreme Fear" | "Fear" | "Neutral" | "Greed" | "Extreme Greed"
  delta1d: number;      // today - yesterday (direction of movement)
  fetchedAt: number;
}

let cached: FearGreedReading | null = null;
let lastFetchAt = 0;
const CACHE_TTL_MS = 30 * 60_000; // 30 min

export async function fetchFearGreed(): Promise<FearGreedReading | null> {
  if (cached && Date.now() - lastFetchAt < CACHE_TTL_MS) return cached;

  try {
    const res = await fetch('https://api.alternative.me/fng/?limit=2', {
      signal: AbortSignal.timeout(5_000),
    });
    if (!res.ok) {
      log('warn', `Fear & Greed fetch failed: ${res.status}`);
      return cached;
    }
    const data = await res.json() as FGIResponse;
    if (!data.data || data.data.length < 1) return cached;

    const today = parseInt(data.data[0]!.value);
    const yesterday = data.data.length > 1 ? parseInt(data.data[1]!.value) : today;

    cached = {
      index: today,
      label: data.data[0]!.value_classification,
      delta1d: today - yesterday,
      fetchedAt: Date.now(),
    };
    lastFetchAt = Date.now();
    return cached;
  } catch (err) {
    log('warn', `Fear & Greed network error: ${String(err)}`);
    return cached;
  }
}

export function fearGreedToMarketPhase(fgi: number): MarketPhase {
  if (fgi <= 20) return 'extreme_fear';
  if (fgi >= 80) return 'extreme_greed';
  if (fgi <= 40) return 'bear';
  if (fgi >= 60) return 'bull';
  return 'neutral';
}

export function buildMarketContext(fgi: FearGreedReading, btcDominance: number): MarketContext {
  return {
    phase: fearGreedToMarketPhase(fgi.index),
    btcDominance,
    fearGreedIndex: fgi.index,
    totalMarketCapChangeD1: 0, // populated separately from CoinGecko
    timestamp: Date.now(),
  };
}

File: src/signals/fear_greed.py
"""Fear & Greed Index fetcher (Alternative.me)."""

import threading
import time
from dataclasses import dataclass
from typing import Optional

import requests

from src.signals._circuit_breaker import CircuitBreaker
from src.storage.database import log
from src.types import MarketContext


@dataclass
class FearGreedReading:
    index: int
    label: str
    delta1d: int
    fetched_at: float


_cached: Optional[FearGreedReading] = None
_last_fetch_at: float = 0
_CACHE_TTL_MS = 30 * 60_000
_breaker = CircuitBreaker("fear_greed")
_lock = threading.Lock()


def fetch_fear_greed() -> Optional[FearGreedReading]:
    global _cached, _last_fetch_at
    now = time.time() * 1000

    with _lock:
        if _cached and now - _last_fetch_at < _CACHE_TTL_MS:
            return _cached
        cached_snapshot = _cached
        last_fetch_snapshot = _last_fetch_at

    # Staleness warning
    if cached_snapshot and last_fetch_snapshot > 0 and now - last_fetch_snapshot > 2 * _CACHE_TTL_MS:
        log("warn", f"Fear & Greed data is stale (last fetch {(now - last_fetch_snapshot) / 60_000:.0f}m ago)")

    if not _breaker.can_call():
        log("warn", "Fear & Greed circuit breaker OPEN — returning cached data")
        return cached_snapshot

    try:
        res = requests.get("https://api.alternative.me/fng/?limit=2", timeout=5)
        if res.status_code != 200:
            log("warn", f"Fear & Greed fetch failed: {res.status_code}")
            _breaker.record_failure()
            return cached_snapshot
        data = res.json()
        items = data.get("data", [])
        if not items:
            _breaker.record_failure()
            return cached_snapshot

        today = int(items[0]["value"])
        yesterday = int(items[1]["value"]) if len(items) > 1 else today

        reading = FearGreedReading(
            index=today,
            label=items[0]["value_classification"],
            delta1d=today - yesterday,
            fetched_at=now,
        )
        with _lock:
            _cached = reading
            _last_fetch_at = now
        _breaker.record_success()
        return reading
    except Exception as err:
        log("warn", f"Fear & Greed network error: {err}")
        _breaker.record_failure()
        return cached_snapshot


def fear_greed_to_market_phase(fgi: int) -> str:
    if fgi <= 20:
        return "extreme_fear"
    if fgi >= 80:
        return "extreme_greed"
    if fgi <= 40:
        return "bear"
    if fgi >= 60:
        return "bull"
    return "neutral"


def build_market_context(fgi: FearGreedReading, btc_dominance: float) -> MarketContext:
    return MarketContext(
        phase=fear_greed_to_market_phase(fgi.index),
        btc_dominance=btc_dominance,
        fear_greed_index=fgi.index,
        total_market_cap_change_d1=0,
        timestamp=time.time() * 1000,
    )
